src/librustc_mir/interpret/operand.rs

```diff
@@ -577,97 +577,125 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+                    Some(discr) => {
+                        // This type actually has discriminants.
+                        let discr_layout = self.layout_of(discr.ty)?;
+                        Scalar::from_uint(discr.val, discr_layout.size)
+                    }
+                    None => {
+                        // On a type without actual discriminants, return variant idx as `u8`.
+                        let discr_layout = self.layout_of(self.tcx.types.u8)?;
+                        Scalar::from_uint(index.as_u32(), discr_layout.size)
+                    }
+                };
+                return Ok((discr, index));
             }
-            Variants::Multiple { discr: ref discr_layout, ref discr_kind, discr_index, .. } => {
-                (discr_layout, discr_kind, discr_index)
+            Variants::Multiple { ref discr, ref discr_kind, discr_index, .. } => {
+                (discr, discr_kind, discr_index)
             }
         };
 
-        // read raw discriminant value
-        let discr_op = self.operand_field(rval, discr_index)?;
-        let discr_val = self.read_immediate(discr_op)?;
-        let raw_discr = discr_val.to_scalar_or_undef();
-        trace!("discr value: {:?}", raw_discr);
-        // post-process
-        Ok(match *discr_kind {
+        // There are *three* types/layouts that come into play here:
+        // - The discriminant has a type for typechecking. This is `discr_ty`, and is used for
+        //   the `Scalar` we return.
+        // - The discriminant gets encoded as a tag/niche, with layout `tag_layout`.
+        //   This is always an integer, and used to interpret the value we read from the
+        //   tag field. For the return value, a cast to `discr_ty` is performed.
+        // - The field storing the tag has a layout, which is very similar to
+        //   `tag_layout` but may be a pointer. This is `tag_val.layout`;
+        //   we just use it for sanity checks.
+
+        // Get layout for tag.
+        let tag_layout = self.layout_of(tag_scalar_layout.value.to_int_ty(*self.tcx))?;
+
+        // Read tag and sanity-check `tag_layout`.
+        let tag_val = self.read_immediate(self.operand_field(rval, tag_index)?)?;
+        assert_eq!(tag_layout.size, tag_val.layout.size);
+        assert_eq!(tag_layout.abi.is_signed(), tag_val.layout.abi.is_signed());
+        let tag_val = tag_val.to_scalar()?;
+        trace!("tag value: {:?}", tag_val);
+
+        // Get type used by typechecking.
+        let discr_ty = match rval.layout.ty.kind {
+            ty::Adt(adt, _) => {
+                let discr_int_ty = Integer::from_attr(self, adt.repr.discr_type());
+                // The signedness of tag and discriminant is the same.
+                discr_int_ty.to_ty(*self.tcx, tag_layout.abi.is_signed())
+            }
+            ty::Generator(_, substs, _) => {
+                let substs = substs.as_generator();
+                substs.discr_ty(*self.tcx)
```

Comment:
So, something is wrong here. I have a generator where `discr_ty` here returns a 32bit type, but then the MIR `Discriminant` statement is used to put this into a 64bit type. IOW, the MIR is ill-formed, as far as I can tell.

[Here's the code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=75c57a7439eb29b9f9ba346ab5d565ca), and here the relevant MIR snippet:
```
fn basic::{{closure}}#0(_1: std::pin::Pin<&mut [generator@src/main.rs:37:15: 37:25 {usize, ()}]>, _2: ()) -> std::ops::GeneratorState<usize, ()> {
    let mut _0: std::ops::GeneratorState<usize, ()>; // return place in scope 0 at src/main.rs:37:15: 37:25
    let mut _3: ();                      // in scope 0 at src/main.rs:37:18: 37:18
    let mut _4: isize;                   // in scope 0 at src/main.rs:37:15: 37:25

    bb0: {
        _4 = discriminant((*(_1.0: &mut [generator@src/main.rs:37:15: 37:25 {usize, ()}]))); // scope 0 at src/main.rs:37:15: 37:25
        switchInt(move _4) -> [0u32: bb1, 1u32: bb3, 3u32: bb2, otherwise: bb4]; // scope 0 at src/main.rs:37:15: 37:25
    }
```
Notice `_4` is ptr-sized, but (if I interpret everything correctly), `discr_ty` for that generator is `u32` -- that's what my logs say, anyway:
```
[TRACE rustc_mir::interpret::operand] tag value: 0x00
[TRACE rustc_mir::interpret::operand] discriminant type: u32
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `Size { raw: 4 }`,
 right: `Size { raw: 8 }`: Size mismatch when writing bits', /home/r/src/rust/rustc.2/src/libstd/macros.rs:16:9
```

Reply:
https://github.com/rust-lang/rust/pull/72502 fixes this.